{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "943c0a74",
   "metadata": {},
   "source": [
    "# ***<font color=\"RoyalBlue\">Execution of NCBI AMRfinderPlus</font>***\n",
    "\n",
    "Detection of antimicrobial resistance genes, pathogenicity-related genes, and other genetic elements.\n",
    "For details on NCBI AMRFinderPlus, please refer to the following website:\n",
    "https://github.com/ncbi/amr\n",
    "\n",
    "<br>\n",
    "<br>\n",
    "\n",
    "---\n",
    "\n",
    "# ***<u><font color=\"RoyalBlue\">Analysis Procedure</font></u>***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e65a294",
   "metadata": {},
   "source": [
    "## 1. Upload Input Files\n",
    "\n",
    "Upload FASTA files (assembled sequences) to the analysis project folder.  \n",
    "You can upload files using the ⬆︎ button at the top of the left sidebar, or by dragging and dropping files into the open analysis project folder.\n",
    "<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "646e76d8",
   "metadata": {},
   "source": [
    "## 2. Create a Strain List\n",
    "\n",
    "Create a strain list using one of the following methods:\n",
    "\n",
    "### 2.1. Create in Jupyter Lab\n",
    "- Click the \"<font color=\"Tomato\">+</font>\" button at the top of the left sidebar → select \"<font color=\"Tomato\">Text file</font>\" from \"<font color=\"Tomato\">Other</font>\" to create a new file.\n",
    "- After editing is complete, press Ctrl+s (Mac: Command+s) to save the file with a desired filename.\n",
    "### 2.2. Create using Notepad, Text Editor, Vim, etc., and copy to the analysis folder.\n",
    "- If created on Windows, the analysis may not work correctly due to differences in line break codes.\n",
    "- Please create the file on Linux or Mac.\n",
    "### 2.3. Create using the following command.\n",
    "- After \"<font color=\"Tomato\">file_name = </font>\", enter the list filename enclosed in double quotes (\") (example: \"filename\")\n",
    "- After \"<font color=\"Tomato\">user_input = (\"\"\"</font>\", enter the strain list (one strain name per line).\n",
    "\n",
    "<font color=\"Tomato\">────────────── ↓↓↓ ***Execute Command*** ↓↓↓ ──────────────</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca3013fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enter the list filename\n",
    "file_name = \"list_amrfinder.txt\"\n",
    "\n",
    "# Paste the strain list\n",
    "user_input = (\"\"\"\n",
    "A0001\n",
    "A0002\n",
    "A0003\n",
    "A0004\n",
    "A0005\n",
    "A0006\n",
    "A0007\n",
    "A0008\n",
    "\"\"\"\n",
    ")\n",
    "\n",
    "# Remove leading/trailing newlines and add a newline at the end\n",
    "user_input = user_input.strip(\"\\n\") + \"\\n\"\n",
    "\n",
    "# Write the input to a file\n",
    "with open(file_name, \"w\") as file:\n",
    "    file.write(user_input)\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b746f4bf",
   "metadata": {},
   "source": [
    "<font color=\"Tomato\">────────────────────────────────────────────</font>\n",
    "<br>\n",
    "\n",
    "## 3. Execute amrfinder_batch.sh\n",
    "\n",
    "- After \"<font color=\"Tomato\">file_name = </font>\", enter the strain list filename created in the previous step, enclosed in double quotes (\").\n",
    "- If you need to modify any parameters listed under \"<font color=\"Tomato\">Set Parameters</font>\", modify them and execute the command.\n",
    "    - You can also execute the command with the default parameters.\n",
    "\n",
    "<font color=\"Tomato\">────────────── ↓↓↓ ***Execute Command*** ↓↓↓ ──────────────</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eaa8e792",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "####################################################\n",
    "# Set Parameters\n",
    "####################################################\n",
    "# Enter the strain list filename created in step 2\n",
    "file_name = \"list_amrfinder.txt\"\n",
    "# Enter the FASTA file extension\n",
    "file_extension = \".fasta\"\n",
    "# Enter the number of threads\n",
    "threads = \"8\"\n",
    "# To use the plus feature (detects pathogenicity genes, metal resistance genes, etc., in addition to antimicrobial resistance genes),\n",
    "# enter \"1\"; to disable, enter \"0\"\n",
    "plus = \"1\"\n",
    "# To specify a genus/species,\n",
    "# uncomment only one from the following list.\n",
    "# If the genus/species is unknown or not in the list, leave all commented (accuracy may be reduced).\n",
    "# species = \"Acinetobacter_baumannii\"\n",
    "# species = \"Bordetella_pertussis\"\n",
    "# species = \"Burkholderia_cepacia\"\n",
    "# species = \"Burkholderia_mallei\"\n",
    "# species = \"Burkholderia_pseudomallei\"\n",
    "# species = \"Campylobacter\"\n",
    "# species = \"Citrobacter_freundii\"\n",
    "# species = \"Clostridioides_difficile\"\n",
    "# species = \"Corynebacterium_diphtheriae\"\n",
    "# species = \"Enterobacter_asburiae\"\n",
    "# species = \"Enterobacter_cloacae\"\n",
    "# species = \"Enterococcus_faecalis\"\n",
    "# species = \"Enterococcus_faecium\"\n",
    "# species = \"Escherichia\"\n",
    "# species = \"Haemophilus_influenzae\"\n",
    "# species = \"Helicobacter_pylori\"\n",
    "# species = \"Klebsiella_oxytoca\"\n",
    "# species = \"Klebsiella_pneumoniae\"\n",
    "# species = \"Neisseria_gonorrhoeae\"\n",
    "# species = \"Neisseria_meningitidis\"\n",
    "# species = \"Pseudomonas_aeruginosa\"\n",
    "# species = \"Salmonella\"\n",
    "# species = \"Serratia_marcescens\"\n",
    "# species = \"Staphylococcus_aureus\"\n",
    "# species = \"Staphylococcus_pseudintermedius\"\n",
    "# species = \"Streptococcus_agalactiae\"\n",
    "# species = \"Streptococcus_pneumoniae\"\n",
    "# species = \"Streptococcus_pyogenes\"\n",
    "# species = \"Vibrio_cholerae\"\n",
    "# species = \"Vibrio_parahaemolyticus\"\n",
    "# species = \"Vibrio_vulnificus\"\n",
    "\n",
    "\n",
    "####################################################\n",
    "# Run amrfinder_batch.sh\n",
    "# **Do not modify the following**\n",
    "####################################################\n",
    "# Run amrfinder_batch.sh\n",
    "plus_option = f\"--plus \" if plus == \"1\" else \"\"\n",
    "species_option = f\"-s {species} \" if 'species' in locals() else \"\"\n",
    "\n",
    "!bash amrfinder_batch.sh \\\n",
    "    -i $file_name \\\n",
    "    -e $file_extension \\\n",
    "    -t $threads $plus_option$species_option\\\n",
    "    | tee amrfinder_batch.log\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea4b6cac",
   "metadata": {},
   "source": [
    "<font color=\"Tomato\">────────────────────────────────────────────</font>\n",
    "\n",
    "<br>\n",
    "\n",
    "### 4. After Execution\n",
    "\n",
    "After execution, a folder named `amrfinder_[date]_[time]_[list_name]` will be created. The files in the folder are as follows:\n",
    "\n",
    "- [strain_name]_out.tsv: Result file when the \"plus\" option is disabled.\n",
    "- [strain_name]_plus_out.tsv: Result file when the \"plus\" option is enabled.\n",
    "- log folder: Logs are collected in this folder. Please refer to this folder if the analysis does not work as expected."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e749cc8f",
   "metadata": {},
   "source": [
    "## 5. Output Format Description\n",
    "\n",
    "The description of the columns in the result file (TSV format) generated after execution is as follows:\n",
    "\n",
    "| Column Name | Description |\n",
    "|------|------|\n",
    "| **Protein id** | ID obtained from the FASTA header line of the protein or DNA sequence |\n",
    "| **Contig id** | Contig name |\n",
    "| **Start** | Gene start position on the DNA contig (1-based) |\n",
    "| **Stop** | Gene end position on the DNA contig (1-based) |\n",
    "| **Strand** | Sequence direction (\"+\" = forward strand, \"-\" = reverse strand) |\n",
    "| **Element symbol** | Gene symbol or SNP definition. For point mutations, the gene symbol and mutation definition are joined with \"_\" |\n",
    "| **Element name** | Full name of protein, RNA, or point mutation |\n",
    "| **Scope** | Database scope: \"core\" (resistance effect expected) or \"plus\" (additional genes of interest) |\n",
    "| **Type** | Functional category: \"AMR\" (antimicrobial resistance), \"STRESS\" (stress response), or \"VIRULENCE\" (pathogenicity) |\n",
    "| **Subtype** | Further detailed classification of the type |\n",
    "| **Class** | For AMR genes, the drug class involved in this gene |\n",
    "| **Subclass** | Details within the drug class |\n",
    "| **Method** | Hit type (ALLELE, EXACT, BLAST, PARTIAL, HMM, INTERNAL_STOP, POINT, etc.) |\n",
    "| **Target length** | Length of query protein or gene (amino acids or nucleotides) |\n",
    "| **Reference sequence length** | Length of the reference sequence in the database (if BLAST hit detected) |\n",
    "| **% Coverage of reference** | Percentage of reference covered by BLAST hit (%) |\n",
    "| **% Identity to reference** | Similarity to reference protein or nucleotide sequence (%) |\n",
    "| **Alignment length** | Length of BLAST alignment (amino acids or nucleotides) |\n",
    "| **Closest reference accession** | RefSeq accession number of the BLAST hit |\n",
    "| **Closest reference name** | Full name of the reference sequence of the BLAST hit |\n",
    "| **HMM accession** | HMM accession number (for HMM search), NA otherwise |\n",
    "| **HMM description** | Family name associated with HMM (for HMM search), NA otherwise |"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
