{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7b6b0d92",
   "metadata": {},
   "source": [
    "# ***<font color=\"RoyalBlue\">Assembly and Data QC (grape_qc_assembly)</font>***\n",
    "\n",
    "Performs coverage calculation, assembly (SPAdes or SKESA), and quality check (QUAST, checkM).\n",
    "\n",
    "<br>\n",
    "<br>\n",
    "\n",
    "---\n",
    "\n",
    "# ***<u><font color=\"RoyalBlue\">Analysis Steps</font></u>***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21543777",
   "metadata": {},
   "source": [
    "## 1. Upload Input Files\n",
    "\n",
    "Upload FASTQ files to your analysis project folder.\n",
    "You can upload files from the ⬆︎ button at the top of the left sidebar, or by dragging and dropping files into the analysis project folder.\n",
    "<br>\n",
    "\n",
    "\n",
    "The file extension must be `fastq.gz`.<br>\n",
    "The file name should begin with the strain name.<br>\n",
    "For example, the following formats are supported:\n",
    "\n",
    "- Illumina format (e.g., strain_S00_L001_R1_001.fastq.gz)\n",
    "- Simple format (e.g., strain_R1.fastq.gz)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe5ab33b",
   "metadata": {},
   "source": [
    "## 2. Create a Strain List\n",
    "\n",
    "Create a strain list using one of the following methods:\n",
    "\n",
    "### 2.1. Create in Jupyter Lab\n",
    "- Click the \"+ \" button at the top of the left sidebar → Click \"Text file\" under \"Other\" to create a new file.\n",
    "- After editing is complete, press Ctrl+s (Mac: Command+s) to save the file with a name.\n",
    "### 2.2. Create with Notepad, Text Editor, Vim, etc. (any text editor) and copy it to the analysis folder.\n",
    "- Creating on Windows may not work correctly (due to line ending differences).\n",
    "- Create on Linux or Mac.\n",
    "### 2.3. Create using the command below.\n",
    "- After \"file_name = \", enter the list name enclosed in double quotes (e.g., \"filename\").\n",
    "- After \"user_input = (\"\"\", enter the strain list (one strain name per line).\n",
    "\n",
    "<font color=\"Tomato\">────────────── ↓↓↓ ***Execute Command*** ↓↓↓ ──────────────</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eecc5ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enter the filename for the list\n",
    "file_name = \"list.txt\"\n",
    "\n",
    "# Paste the strain list\n",
    "user_input = (\"\"\"\n",
    "A0001\n",
    "A0002\n",
    "A0003\n",
    "A0004\n",
    "A0005\n",
    "A0006\n",
    "A0007\n",
    "A0008\n",
    "\"\"\"\n",
    ")\n",
    "\n",
    "####################################################\n",
    "# **Do not modify the following**\n",
    "####################################################\n",
    "\n",
    "# Remove leading and trailing newlines, then add a newline\n",
    "user_input = user_input.strip(\"\\n\") + \"\\n\"\n",
    "\n",
    "# Write the input elements to the file\n",
    "with open(file_name, \"w\") as file:\n",
    "    file.write(user_input)\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9e41ae5",
   "metadata": {},
   "source": [
    "<font color=\"Tomato\">────────────────────────────────────────────</font>\n",
    "<br>\n",
    "\n",
    "## 3. Create a FASTQ List\n",
    "\n",
    "Create a list of FASTQ file names (short-read data file names) associated with each strain in the strain list created in step 2. Follow the procedure below.\n",
    "\n",
    "### 3.1. Run find_strain_pairs.py\n",
    "- After \"file_name = \", enter the strain list name created in the previous step, enclosed in double quotes.\n",
    "- In \"Set Parameters\" section, modify any items if needed and execute. You can also execute with the content as shown.\n",
    "\n",
    "<font color=\"Tomato\">────────────── ↓↓↓ ***Execute Command*** ↓↓↓ ──────────────</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d33436a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "####################################################\n",
    "# Set Parameters\n",
    "####################################################\n",
    "# Enter the filename of the strain list created in step 2\n",
    "file_name = \"list.txt\"\n",
    "# Enter the file extension for FASTQ files\n",
    "file_extension = \"fastq.gz\"\n",
    "# Filename for the FASTQ list to be created\n",
    "fastq_list_name = \"list_fastq.tsv\"\n",
    "# Filename for the list of strains with no paired FASTQ files found\n",
    "unpaired_list = \"unpaired_fastq.tsv\"\n",
    "\n",
    "####################################################\n",
    "# Run find_strain_pairs.py\n",
    "# **Do not modify the following**\n",
    "####################################################\n",
    "command = [\n",
    "    \"find_strain_pairs.py\", file_name,\n",
    "    \"--file_extension\", file_extension,\n",
    "    \"--paired_list\", fastq_list_name,\n",
    "    \"--unpaired_list\", unpaired_list\n",
    "]\n",
    "subprocess.run(command, capture_output=False, text=True)\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "183426e8",
   "metadata": {},
   "source": [
    "<font color=\"Tomato\">────────────────────────────────────────────</font>\n",
    "<br>\n",
    "\n",
    "### 3.2. Verify and Modify the FASTQ List\n",
    "- Open and verify `list_fastq.tsv` (FASTQ list) to confirm that each strain name has two FASTQ files (Read1 and Read2) correctly listed.\n",
    "- If `unpaired_fastq.tsv` is generated, open and verify it to identify strains where FASTQ file pairs were not found.\n",
    "    - Copy the strain name to `list_fastq.tsv` and add the paired FASTQ files separated by tabs.\n",
    "- Below is an example of the FASTQ list format (columns are tab-separated, line endings are \\n)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25348a33",
   "metadata": {},
   "source": [
    "```plaintxt\n",
    "strain1\tstrain1_R1.fastq.gz\tstrain1_R2.fastq.gz\n",
    "strain2\tstrain2_R1.fastq.gz\tstrain2_R2.fastq.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c54774bc",
   "metadata": {},
   "source": [
    "## 4. Create a Configuration File<br>\n",
    "\n",
    "<b>If QC is not necessary, Proceed to Step 5.</b><br>\n",
    "<br>\n",
    "\n",
    "If you want to perform QC, create a configuration.txt file as follows:<br>\n",
    "\n",
    "There is a file called `stec/configuration.txt` in the `sample_data` folder within the folder where the `CreateProject.ipynb` file used in Create Project is located. This is a sample configuration for STEC. Copy it to your project folder and edit it as appropriate for the bacterial species you are analyzing.<br>\n",
    "\n",
    "In configuration.txt, you can specify genome size and required coverage to determine if the data amount is sufficient.<br>\n",
    "Genome size varies by bacterial species, so you need to specify it appropriately.<br>\n",
    "Items that do not require checking can be deleted, and no check will be performed on them.<br>\n",
    "<br>\n",
    "Below is an example for STEC.<br>\n",
    "<br>\n",
    "Genome size (Mbp):&nbsp;&nbsp;5.5 <br>\n",
    "Coverage threshold:&nbsp;&nbsp;40 <br>\n",
    "Maximum number of contigs:&nbsp;&nbsp;1000 <br>\n",
    "Minimum genome size (Mbp):&nbsp;&nbsp;4 <br>\n",
    "Maximum genome size (Mbp):&nbsp;&nbsp;6.5 <br>\n",
    "Minimum completeness (%):&nbsp;&nbsp;98 <br>\n",
    "Maximum contamination (%):&nbsp;&nbsp;2 <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51ea5b4b",
   "metadata": {},
   "source": [
    "## 5. Run qc_assembly (Data Check and Assembly)\n",
    "\n",
    "Depending on your environment, choose whether to run qc_assembly in 5.1 or 5.2.\n",
    "\n",
    "### 5.1. When Machine Memory is 40GB or More\n",
    "\n",
    "Change the list part in the cell below (the part enclosed in double quotes in \"list = \\\"list_fastq.tsv\\\"\") to the name of the list you created and execute it.\n",
    "\n",
    "|Variable|Parameter|Required|Description|\n",
    "|---|---|---|---|\n",
    "|list|-i|●|The filename of the FASTQ list created in step 4.|\n",
    "|assembler|-a|-|Assembler to use [p:SPAdes, k:SKESA] (p by default)|\n",
    "|config|-c|-|Configuration file for quality thresholds during grape analysis. If not specified, quality assessment will not be performed.|\n",
    "|cleanup|-u|-|Whether to delete intermediate files. [1: cleanup, 0: not cleanup] (1 by default)|\n",
    "|cutoff length|-L|-|Remove contigs shorter than the specified length. (500 by default)|\n",
    "|fastp|-p|-|Whether to perform quality trimming with fastp (remove low-quality bases). [1:run, 0:not run] (1 by default)|\n",
    "|scaffold|-s|-|Whether to create scaffold sequences. Only effective when SPAdes is specified as the assembler. [1: create, 0: not create] (0 by default)|\n",
    "|threads|-t|-|Number of threads (8 by default)|\n",
    "\n",
    "<br>\n",
    "\n",
    "<font color=\"Tomato\">────────────── ↓↓↓ ***Execute Command*** ↓↓↓ ──────────────</font><br>\n",
    "\\* For log information, refer to `grape_qc_assembly.log` in the execution directory. （<u><font color=\"Red\">***This file is overwritten each time you run.***</u></font>）"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d3a6762",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################\n",
    "# Set Parameters\n",
    "####################################################\n",
    "# Strain list\n",
    "list = \"list_fastq.tsv\"\n",
    "# Configuration file\n",
    "# If you want to perform quality assessment, create a quality configuration file and specify its filename below.\n",
    "config = \"\" \n",
    "# config = \"configuration.txt\"\n",
    "# Assembler: Enter \"p\" to use SPAdes, \"k\" to use SKESA\n",
    "assembler = \"p\"\n",
    "# Contig threshold: Contigs shorter than this value (bp) will be removed\n",
    "cutoff = 500\n",
    "# Enter \"1\" to perform trimming with fastp, \"0\" to skip\n",
    "fastp = 1\n",
    "# Enter \"1\" to create scaffolds, \"0\" to skip\n",
    "scaffold = 0\n",
    "# Number of threads\n",
    "threads = 8\n",
    "# Enter \"0\" to keep intermediate files like trimmed fastq, \"1\" to delete them\n",
    "cleanup = 1\n",
    "\n",
    "####################################################\n",
    "# Run grape_qc_assembly (Do not modify)\n",
    "####################################################\n",
    "# Run grape_qc_assembly\n",
    "config_option = f\"-c {config} \" if config else \"\"\n",
    "\n",
    "!bash grape_qc_assembly.sh \\\n",
    "    -i $list \\\n",
    "    -a $assembler \\\n",
    "    -L $cutoff \\\n",
    "    -p $fastp \\\n",
    "    -s $scaffold \\\n",
    "    -t $threads \\\n",
    "    -u $cleanup $config_option\\\n",
    "    > grape_qc_assembly.log\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f099c93",
   "metadata": {},
   "source": [
    "### 5.2. When Machine Memory is 40GB or Less\n",
    "\n",
    "Change the list part in the cell below (the part enclosed in double quotes in \"list = \\\"list_fastq.tsv\\\"\") to the name of the list you created and execute it.<br>\n",
    "\\* For details on each parameter, see `5.1. When Machine Memory is 40GB or More`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be953542",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################\n",
    "# Set Parameters\n",
    "####################################################\n",
    "# Strain list\n",
    "list = \"list_fastq.tsv\"\n",
    "# Configuration file\n",
    "# If you want to perform quality assessment, create a quality configuration file and specify its filename below.\n",
    "config = \"\" \n",
    "# config = \"configuration.txt\"\n",
    "# Assembler: Enter \"p\" to use SPAdes, \"k\" to use SKESA\n",
    "assembler = \"p\"\n",
    "# Contig threshold: Contigs shorter than this value (bp) will be removed\n",
    "cutoff = 500\n",
    "# Enter \"1\" to perform trimming with fastp, \"0\" to skip\n",
    "fastp = 1\n",
    "# Enter \"1\" to create scaffolds, \"0\" to skip\n",
    "scaffold = 0\n",
    "# Number of threads\n",
    "threads = 8\n",
    "# Enter \"0\" to keep intermediate files like trimmed fastq, \"1\" to delete them\n",
    "cleanup = 1\n",
    "\n",
    "####################################################\n",
    "# Run grape_qc_assembly (Do not modify)\n",
    "####################################################\n",
    "# Run grape_qc_assembly\n",
    "config_option = f\"-c {config} \" if config else \"\"\n",
    "\n",
    "!bash grape_qc_assembly.sh \\\n",
    "    -i $list \\\n",
    "    -a $assembler \\\n",
    "    -L $cutoff \\\n",
    "    -p $fastp \\\n",
    "    -s $scaffold \\\n",
    "    -t $threads \\\n",
    "    -u $cleanup $config_option \\\n",
    "    -r 1 \\\n",
    "    > grape_qc_assembly.log\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6938437",
   "metadata": {},
   "source": [
    "<font color=\"Tomato\">────────────────────────────────────────────</font>\n",
    "\n",
    "<br>\n",
    "\n",
    "### 5. After execution, a folder named `qc_assembly_[date]_[time]_[list_name]` will be created. The files in the folder are as follows:<br>\n",
    "\n",
    "- [strain_name].fasta: Draft genome\n",
    "- assembly_summary.tsv: Summary of statistics such as contig lengths (calculated by QUAST).\n",
    "- coverage.txt: Coverage list\n",
    "- list_assembly.tsv: FASTQ file list for strains included in list_over_coverage.\n",
    "- list_over_coverage: List of strains that exceeded the coverage specified in the configuration file. Only strains in this list will be assembled.\n",
    "- qc_results.tsv: Results from QUAST and CheckM (genome size, contamination, etc.). Records \"pass\" if within the configuration file range, \"fail\" if outside. The qc_results (overall assessment) column shows \"PASS\" only when all values pass.\n",
    "- qc_results.xlsx: Excel version of qc_results.tsv (same content).\n",
    "- Data only retained when cleanup=0\n",
    "  - fastp folder: FASTQ files trimmed by fastp\n",
    "    - [strain_name]_1(2).fastq.gz: Read1 or 2 read files where pairs were found\n",
    "    - [strain_name]_u1(u2).fastq.gz: Read1 or 2 read files where pairs were not found\n",
    "  - quast folder: QUAST execution results\n",
    "    - For details of output, see [https://github.com/ablab/quast#output](https://github.com/ablab/quast#output).\n",
    "  - checkm folder: CheckM execution results\n",
    "    - For details of output, see [https://github.com/Ecogenomics/CheckM](https://github.com/Ecogenomics/CheckM).\n",
    "\n",
    "Generally, data that passes QC is used for subsequent analysis (SNP analysis, etc.).<br>\n",
    "However, data that does not pass quality checks may not be completely unusable. If there is no obvious contamination or similar issues, examine the data carefully."
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
