{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "871403f0",
   "metadata": {},
   "source": [
    "# ***<font color=\"RoyalBlue\">SNP Analysis Additional Processing (SNPclipper/SNPcaster_gubbins) Notebook</font>***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19960cf2",
   "metadata": {},
   "source": [
    "<br>\n",
    "<br>\n",
    "\n",
    "***\n",
    "\n",
    "# ***<u><font color=\"RoyalBlue\">Analysis Procedure</font></u>***\n",
    "\n",
    "<font color=\"DarkOrange\">\\* This notebook explains how to perform additional analyses on the results of a previously run SNPcaster by changing conditions.</font></br>\n",
    "<font color=\"DarkOrange\">\\* If you have not yet run SNPcaster, please execute <b>SNPcaster_quickstart.ipynb</b> or <b>SNPcaster.ipynb</b> first.</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e31deec",
   "metadata": {},
   "source": [
    "## 1. Run SNPcaster\n",
    "\n",
    "Please execute <b>SNPcaster_quickstart.ipynb</b> or <b>SNPcaster.ipynb</b> first if you have not yet run SNPcaster."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afc9f7aa",
   "metadata": {},
   "source": [
    "## 2. Select the program to run\n",
    "\n",
    "From the list below, select the analysis you want to perform and go to the corresponding section.</br>\n",
    "#### - To see results after removing recombinogenic regions detected by Gubbins → <font color=\"Lime\"> <b>Go to 3</b></font><br>\n",
    "#### - To see results when changing the conditions of clustered SNPs or masking → <font color=\"Lime\"><b>Go to 4</b></font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2db61ef5",
   "metadata": {},
   "source": [
    "## 3. Additional execution of Gubbins\n",
    "\n",
    "### 3-1. How to run\n",
    "\n",
    "You can run Gubbins by specifying the SNPcaster output folder that has not yet been processed by Gubbins.\n",
    "\n",
    "#### Options\n",
    "\n",
    "    *Optional. If not specified, default values will be used.\n",
    "\n",
    "|Parameter|Description|\n",
    "|---|---|\n",
    "|-i|Target SNPcaster output folder for additional Gubbins execution|\n",
    "|-t*|Number of threads (default is 8)|\n",
    "\n",
    "<font color=\"Tomato\">────────────── 　↓↓↓ 　***Command Execution***　 ↓↓↓　 ──────────────</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b219f779",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################\n",
    "# Set parameters\n",
    "####################################################\n",
    "# SNPcaster output folder\n",
    "input ='snpcaster_20240115_094704_list_test'\n",
    "# Number of threads\n",
    "threads = 8\n",
    "\n",
    "####################################################\n",
    "# Run snpcaster_gubbins.sh\n",
    "####################################################\n",
    "!bash snpcaster_gubbins.sh -i $input -t $threads\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e740ed53",
   "metadata": {},
   "source": [
    "<font color=\"Tomato\">────────────────────────────────────────────</font>\n",
    "<br>\n",
    "\n",
    "### 3-2. Output\n",
    "\n",
    "The specified folder will contain the Gubbins-processed results: <b>4_results_gubbins</b> and <b>5_results_with_gubbins</b>.<br>\n",
    "\\* For details of the output files, see <b>SNPcaster.ipynb section 9.2</b>.<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5d82632",
   "metadata": {},
   "source": [
    "## 4. Run SNPclipper\n",
    "\n",
    "SNPclipper generates new results from a previously run SNPcaster when you change the conditions of clustered SNPs or masking.\n",
    "<br>\n",
    "Specifically, it regenerates <b>3_results_without_gubbins</b> and <b>5_results_with_gubbins</b>.\n",
    "\n",
    "### 4-1. How to run\n",
    "\n",
    "You can run SNPclipper by specifying the SNPcaster output folder and the conditions for clustered SNPs or masking.\n",
    "\n",
    "#### Options\n",
    "\n",
    "    *Optional. If not specified, default values will be used.\n",
    "    * \"clustered SNP\" refers to regions of recombination that are removed from SNPs.\n",
    "\n",
    "|Parameter|Description|\n",
    "|---|---|\n",
    "|-i|SNPcaster output folder|\n",
    "|-c*|Interval of adjacent SNPs (cluster SNP) to delete (0 means no deletion, default is 0)|\n",
    "|-d*|Mask file (if not specified, no masking is performed)|\n",
    "|-t*|Number of threads (default is 8)|\n",
    "\n",
    "<font color=\"Tomato\">────────────── 　↓↓↓ 　***Command Execution***　 ↓↓↓　 ──────────────</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2d7779b",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################\n",
    "# Set parameters\n",
    "####################################################\n",
    "# SNPcaster output folder\n",
    "input ='snpcaster_20240115_094704_list_test'\n",
    "# Interval of adjacent SNPs (clustered SNP) to delete\n",
    "cluster = 0\n",
    "# Number of threads\n",
    "threads = 8\n",
    "# If you want to apply a mask, remove the leading # and insert the file path between \"\"\n",
    "#mask = \"mask.tsv\"\n",
    "\n",
    "####################################################\n",
    "# Run SNPclipper (do not modify the below)\n",
    "####################################################\n",
    "extra_options = \"\"\n",
    "extra_options += f\"-d {mask} \" if 'mask' in locals() and mask else \"\"\n",
    "\n",
    "!bash snpclipper.sh \\\n",
    "    -i $input \\\n",
    "    -c $cluster \\\n",
    "    -t $threads $extra_options\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "267ca150",
   "metadata": {},
   "source": [
    "<font color=\"Tomato\">────────────────────────────────────────────</font>\n",
    "<br>\n",
    "\n",
    "### 4-2. Output\n",
    "\n",
    "A folder named <b>snpclipper_[execution datetime]</b> will be created in the specified folder,\n",
    "and it will contain the results <b>3_results_without_gubbins</b> and <b>5_results_with_gubbins</b> processed under the specified conditions.<br>\n",
    "\\* <b>5_results_with_gubbins</b> is generated only if Gubbins was also executed.</br>\n",
    "\\* If you also want to run Gubbins, perform <b>3. Additional Gubbins execution</b>.</br>\n",
    "\\* For details of the output files, see <b>SNPcaster.ipynb section 9.2</b>.<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a12ebea",
   "metadata": {},
   "source": [
    "## 5. Analyses using SNPs\n",
    "\n",
    "### 5.1. Creating a Haplotype Network Diagram\n",
    "\n",
    "Open one of the following files with PopART for analysis.\n",
    "<br>\n",
    "- 3_results_without_gubbins\n",
    "    - final_snp.nex\n",
    "    - final_snp_woRef.nex\n",
    "- 5_results_with_gubbins\n",
    "    - final_snp_with_gubbins.nex\n",
    "    - final_snp_with_gubbins_woRef.nex\n",
    "<br>\n",
    "\\* Use the files with \"woRef\" when you want a diagram without the reference sequence.\n",
    "<br>\n",
    "\n",
    "### 5.2. Building a Phylogenetic Tree\n",
    "\n",
    "Use one of the following files (the 5 set is generated only when Gubbins is executed).\n",
    "- 3_results_without_gubbins\n",
    "    - final_snp.fasta\n",
    "    - final_snp_woRef.fasta\n",
    "- 5_results_with_gubbins\n",
    "    - final_snp_with_gubbins.fasta\n",
    "    - final_snp_with_gubbins_woRef.fasta\n",
    "\n",
    "#### Use RAxML-NG → <font color=\"Lime\"><b>Go to 6.1.1.</b></font>\n",
    "#### Use IQTREE → <font color=\"Lime\"><b>Go to 6.1.2.</b></font>\n",
    "\n",
    "### 5.3. Calculating SNPs Between Strains\n",
    "\n",
    "The number of SNPs between strains is listed in the following files and can be opened with Notepad, Excel, etc.\n",
    "- 3_results_without_gubbins\n",
    "    - dist_final_snp.tsv\n",
    "    - dist_final_snp_matrix.tsv \n",
    "- 5_results_with_gubbins\n",
    "    - dist_final_snp_without_recombination.tsv\n",
    "    - dist_final_snp_matrix_without_recombination.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "462e892c",
   "metadata": {},
   "source": [
    "## 6. Building a Phylogenetic Tree\n",
    "\n",
    "Using the obtained SNP files, construct a maximum-likelihood phylogenetic tree with either RAxML-NG or IQTREE.<br>\n",
    "* Change the name of the reference sequence (default is Ref) if necessary.\n",
    "* When building a tree without the reference sequence, use the files with \"woRef\".\n",
    "\n",
    "### 6.1. Maximum-Likelihood Tree Construction\n",
    "\n",
    "\\* Both RAxML-NG and IQTREE can be used."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d14226e",
   "metadata": {},
   "source": [
    "#### 6.1.1 RAxML-NG\n",
    "\n",
    "ModelTest-NG selects the optimal nucleotide substitution model, which is then used by RAxML-NG to build the maximum-likelihood tree.\n",
    "\n",
    "- In the cell below, enter the path to the SNP file after \"input=\".\n",
    "- Example: 'snpcaster_20240115_094704_list.txt/3_results_without_gubbins/final_snp.fasta'<br>\n",
    "    ↑<u><font color=\"Red\">The \"20240115_094704_list.txt\" part varies with each analysis.</font></u>\n",
    "\n",
    "- Options\n",
    "\n",
    "|Parameter|Required|Description|\n",
    "|---|---|---|\n",
    "|input|●|Input file (e.g., xxx.fasta).|\n",
    "|threads|●|Number of threads for modeltest-ng (e.g., 12). RAxML-NG selects threads automatically.|\n",
    "|bootstrap|-|Bootstrap (default 1,000)|\n",
    "\n",
    "<font color=\"Tomato\">────────────── 　↓↓↓ 　***Command Execution***　 ↓↓↓　 ──────────────</font><br>\n",
    "\\* Log can be found in `raxml-ng.log` in the execution directory. (It is overwritten each run.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fd3f3ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################\n",
    "# Set parameters\n",
    "####################################################\n",
    "input='snpcaster_20240115_094704_list.txt/3_results_without_gubbins/final_snp.fasta'\n",
    "# Uncomment and edit the line below if you want to use the results from a Gubbins run.\n",
    "# input = \"snpcaster_20240115_094704_list.txt/5_results_with_gubbins/final_snp_after_gubbins.fasta\"\n",
    "threads=8\n",
    "bootstrap=1000\n",
    "\n",
    "####################################################\n",
    "# Run raxml-ng\n",
    "####################################################\n",
    "!bash raxml-ng.sh $input $threads $bootstrap > raxml-ng.log\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea7b3d3f",
   "metadata": {},
   "source": [
    "<font color=\"Tomato\">────────────────────────────────────────────</font>\n",
    "\n",
    "#### 6.1.2 IQTREE Execution\n",
    "\n",
    "IQTREE automatically selects the nucleotide substitution model and builds a maximum-likelihood tree using the optimal model.\n",
    "\n",
    "- In the cell below, enter the path to the SNP file after \"input=\".\n",
    "- Example: 'snpcaster_20240115_094704_list.txt/3_results_without_gubbins/final_snp.fasta'<br>\n",
    "    ↑<u><font color=\"Red\">The \"20240115_094704_list.txt\" part varies with each analysis.</font></u>\n",
    "\n",
    "- Options\n",
    "\n",
    "|Variable|Required|Description|\n",
    "|---|---|---|\n",
    "|input|●|Input file (e.g., xxx.fasta).|\n",
    "|bootstrap|-|Ultrafast bootstrap replicates (≥1,000, default 1,000)|\n",
    "\n",
    "- Output\n",
    "  The generated iqtree folder (`iqtree_results_<date>_<time>`) will contain the results.\n",
    "\n",
    "<font color=\"Tomato\">────────────── 　↓↓↓ 　***Command Execution***　 ↓↓↓　 ──────────────</font><br>\n",
    "\\* Verify that the `input` variable points to a valid file path.</br>\n",
    "\\* Log can be found in `iqtree.log` in the execution directory. (It is overwritten each run.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c31da4c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################\n",
    "# Set parameters\n",
    "####################################################\n",
    "input='snpcaster_20240115_094704_list.txt/3_results_without_gubbins/final_snp.fasta'\n",
    "# Uncomment and edit the line below if you want to use the results from a Gubbins run.\n",
    "# input = \"snpcaster_20240115_094704_list.txt/5_results_with_gubbins/final_snp_after_gubbins.fasta\"\n",
    "bootstrap=1000\n",
    "\n",
    "####################################################\n",
    "# Run iqtree\n",
    "####################################################\n",
    "!bash iqtree.sh $input $bootstrap > iqtree.log\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa76e63e",
   "metadata": {},
   "source": [
    "<font color=\"Tomato\">────────────────────────────────────────────</font>\n",
    "\n",
    "### 6.2. Output location for phylogenetic tree files\n",
    "\n",
    "**The analysis folder** will contain the following tree result folders and files:\n",
    "\n",
    "- RAxML-NG:\n",
    "  - Folder name: `raxml_results_<date>_<time>`\n",
    "  - File: `final_snp_without_recombination.fasta_bootstrap.nwk` (bootstrap-annotated tree)\n",
    "- IQTREE:\n",
    "  - Folder name: `iqtree_results_<date>_<time>`\n",
    "  - File: `final_snp_without_recombination.fasta.contree` (bootstrap-annotated tree)\n",
    "\n",
    "### 6.3. Verifying phylogenetic tree files\n",
    "\n",
    "Open the tree files with MEGA, CLC Genomics Workbench, etc.<br>\n",
    "In MEGA, you can save images via Image → Copy to Clipboard. To save images with the root, select the root line, click \"Place Root on Branch\", and use \"Show Subtree Separately\".\n",
    "<br>"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
