{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "24d9d29d",
   "metadata": {},
   "source": [
    "# ***<font color=\"RoyalBlue\">SNP Analysis </font>*** <font color=\"Tomato\"> Quick Start </font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b638134",
   "metadata": {},
   "source": [
    "##### Using _E. coli_ O157:H7 Sakai strain as reference sequence and performing SNP analysis with all default parameters.\n",
    "\n",
    "<br>\n",
    "\n",
    "***\n",
    "\n",
    "# ***<u><font color=\"RoyalBlue\">Analysis Steps</font></u>***\n",
    "\n",
    "<font color=\"DarkOrange\">\\* It is recommended to perform Quality check (grape_qc_assembly) on the data before analysis</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cda2d5e0",
   "metadata": {},
   "source": [
    "## 1. Upload Input Files\n",
    "\n",
    "Upload FASTQ files to your analysis project folder.\n",
    "You can upload files from the ⬆︎ button at the top of the left sidebar, or by dragging and dropping files into the analysis project folder.\n",
    "<br>\n",
    "\n",
    "\n",
    "The file extension must be `fastq.gz`.<br>\n",
    "The file name should begin with the strain name.<br>\n",
    "For example, the following formats are supported:\n",
    "\n",
    "- Illumina format (e.g., strain_S00_L001_R1_001.fastq.gz)\n",
    "- Simple format (e.g., strain_R1.fastq.gz)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1d3e2b6",
   "metadata": {},
   "source": [
    "## 2. Create a Strain List\n",
    "\n",
    "Create a strain list using one of the following methods:\n",
    "\n",
    "### 2.1. Create in Jupyter Lab\n",
    "- Click the \"+ \" button at the top of the left sidebar → Click \"Text file\" under \"Other\" to create a new file.\n",
    "- After editing is complete, press Ctrl+s (Mac: Command+s) to save the file with a name.\n",
    "### 2.2. Create with Notepad, Text Editor, Vim, etc. (any text editor) and copy it to the analysis folder.\n",
    "- Creating on Windows may not work correctly (due to line ending differences).\n",
    "- Create on Linux or Mac.\n",
    "### 2.3. Create using the command below.\n",
    "- After \"file_name = \", enter the list name enclosed in double quotes (e.g., \"filename\").\n",
    "- After \"user_input = (\"\"\", enter the strain list (one strain name per line).\n",
    "\n",
    "<font color=\"Tomato\">────────────── ↓↓↓ ***Execute Command*** ↓↓↓ ──────────────</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db360b21",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enter the filename for the list\n",
    "file_name = \"list.txt\"\n",
    "\n",
    "# Paste the strain list\n",
    "user_input = (\"\"\"\n",
    "A0001\n",
    "A0002\n",
    "A0003\n",
    "A0004\n",
    "A0005\n",
    "A0006\n",
    "A0007\n",
    "A0008\n",
    "\"\"\"\n",
    ")\n",
    "\n",
    "####################################################\n",
    "# **Do not modify the following**\n",
    "####################################################\n",
    "\n",
    "# Remove leading and trailing newlines, then add a newline\n",
    "user_input = user_input.strip(\"\\n\") + \"\\n\"\n",
    "\n",
    "# Write the input elements to the file\n",
    "with open(file_name, \"w\") as file:\n",
    "    file.write(user_input)\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0889d45",
   "metadata": {},
   "source": [
    "<font color=\"Tomato\">────────────────────────────────────────────</font>\n",
    "<br>\n",
    "\n",
    "## 3. Create a FASTQ List\n",
    "\n",
    "Create a list of FASTQ file names (short-read data file names) associated with each strain in the strain list created in step 2. Follow the procedure below.\n",
    "\n",
    "### 3.1. Run find_strain_pairs.py\n",
    "- After \"file_name = \", enter the strain list name created in the previous step, enclosed in double quotes.\n",
    "- In \"Set Parameters\" section, modify any items if needed and execute. You can also execute with the content as shown.\n",
    "\n",
    "<font color=\"Tomato\">────────────── ↓↓↓ ***Execute Command*** ↓↓↓ ──────────────</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6643061",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "####################################################\n",
    "# Set Parameters\n",
    "####################################################\n",
    "# Enter the filename of the strain list created in step 2\n",
    "file_name = \"list.txt\"\n",
    "# Enter the file extension for FASTQ files\n",
    "file_extension = \"fastq.gz\"\n",
    "# Filename for the FASTQ list to be created\n",
    "fastq_list_name = \"list_fastq.tsv\"\n",
    "# Filename for the list of strains with no paired FASTQ files found\n",
    "unpaired_list = \"unpaired_fastq.tsv\"\n",
    "\n",
    "####################################################\n",
    "# Run find_strain_pairs.py\n",
    "# **Do not modify the following**\n",
    "####################################################\n",
    "command = [\n",
    "    \"find_strain_pairs.py\", file_name,\n",
    "    \"--file_extension\", file_extension,\n",
    "    \"--paired_list\", fastq_list_name,\n",
    "    \"--unpaired_list\", unpaired_list\n",
    "]\n",
    "subprocess.run(command, capture_output=False, text=True)\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4634eb33",
   "metadata": {},
   "source": [
    "<font color=\"Tomato\">────────────────────────────────────────────</font>\n",
    "<br>\n",
    "\n",
    "### 3.2. Verify and Modify the FASTQ List\n",
    "- Open and verify `list_fastq.tsv` (FASTQ list) to confirm that each strain name has two FASTQ files (Read1 and Read2) correctly listed.\n",
    "- Open and verify `unpaired_fastq.tsv` to identify strains where FASTQ file pairs were not found.\n",
    "    - If any are found, copy the strain name to `list_fastq.tsv` and add the paired FASTQ files separated by tabs.\n",
    "- Below is an example of the FASTQ list format (columns are tab-separated, line endings are \\n)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97a59fc4",
   "metadata": {},
   "source": [
    "```plaintxt\n",
    "strain1\tstrain1_R1.fastq.gz\tstrain1_R2.fastq.gz\n",
    "strain2\tstrain2_R1.fastq.gz\tstrain2_R2.fastq.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5eed2a15",
   "metadata": {},
   "source": [
    "## 4. Run SNPcaster\n",
    "<font color=\"DarkOrange\">\\* If you have already run SNPcaster on some strains, move the SNP folder (strain name folder) to the execution folder. SNP analysis will be skipped for those strains, and BactSNP will only be executed for strains without SNP folders.<br>\n",
    "</font>\n",
    "\n",
    "<br>\n",
    "\n",
    "Change the list part in the cell below (the part enclosed in double quotes in \"list = \\\"list.txt\\\"\") to the name of the list you created and execute it.\n",
    "\n",
    "#### Options\n",
    "    *Optional. If not specified, default values will be used.\n",
    "    *Clustered SNPs are regions removed from SNPs as recombination regions.\n",
    "\n",
    "|Parameter|Description|\n",
    "|---|---|\n",
    "|-r|Reference sequence filename (e.g., ref.fasta)|\n",
    "|-i|Strain list (e.g., list)|\n",
    "|-a*|Value of --allele_freq in BactSNP (default is 0.9)|\n",
    "|-c*|Interval for clustered SNPs to be removed (specify 0 to not remove; default is 0)|\n",
    "|-d*|Mask file (no mask processing if not specified)|\n",
    "|-f*|Path to fastq list (required if there are strains to run BactSNP on)|\n",
    "|-g*|Whether to run gubbins (1: run, 0: don't run) (default is 0)|\n",
    "|-j*|Number of BactSNP jobs (default is 4)|\n",
    "|-t*|Number of threads (default is 8)|\n",
    "\n",
    "#### File Format\n",
    "The fastq list should be a tsv (tab-separated) file with strain names and their associated Read1 and Read2 file names as follows:\n",
    "```\n",
    "strain1 strain1_R1.fastq.gz strain1_R2.fastq.gz\n",
    "strain2 strain2_R1.fastq.gz strain2_R2.fastq.gz\n",
    "```\n",
    "\n",
    "<br>\n",
    "\n",
    "<font color=\"Tomato\">────────────── ↓↓↓ ***Execute Command*** ↓↓↓ ──────────────</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81bde63e",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################\n",
    "# Set Parameters\n",
    "####################################################\n",
    "# Reference sequence file\n",
    "reference = '../../sample_data/ehec/Sakai_BA000007.fasta'\n",
    "# Strain list file\n",
    "list = \"list.txt\"\n",
    "# BactSNP allele frequency (--allele_freq value)\n",
    "allele_freq = 0.9\n",
    "# Interval for clustered SNPs to be removed\n",
    "cluster = 0\n",
    "# Run gubbins: 1=yes, 0=no\n",
    "gubbins = 0\n",
    "# Number of BactSNP jobs\n",
    "jobs = 4\n",
    "# Number of threads\n",
    "threads = 8\n",
    "# If you don't want to use the fastq list, add # before fastq_list below\n",
    "fastq_list = \"list_fastq.tsv\"\n",
    "# To perform mask processing, remove the # at the beginning of the line and enter the file path between the quotes\n",
    "#mask = \"../../sample_data/ehec/mask.tsv\"\n",
    "\n",
    "####################################################\n",
    "# Run SNPcaster\n",
    "####################################################\n",
    "\n",
    "extra_options = \"\"\n",
    "extra_options += f\"-f {fastq_list} \" if 'fastq_list' in locals() and fastq_list else \"\"\n",
    "extra_options += f\"-d {mask} \" if 'mask' in locals() and mask else \"\"\n",
    "\n",
    "!bash snpcaster.sh \\\n",
    "    -r $reference \\\n",
    "    -i $list \\\n",
    "    -a $allele_freq \\\n",
    "    -c $cluster \\\n",
    "    -g $gubbins \\\n",
    "    -j $jobs \\\n",
    "    -t $threads $extra_options\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6630d28",
   "metadata": {},
   "source": [
    "<font color=\"Tomato\">────────────────────────────────────────────</font>\n",
    "\n",
    "\n",
    "#### About Output Files\n",
    "\n",
    "After execution, a folder named snpcaster\\_[date]\\_[time]\\_[list_name] will be created.<br>\n",
    "\n",
    "***<font color=\"Red\">\\*\n",
    "    If FASTQ files do not exist, an error will occur and processing will terminate.<br>\n",
    "    A \"missing_list\" will be output as a list of strains without FASTQ files.<br></font>***\n",
    "\n",
    "For information about output files and folders, and precautions, refer to the \"9.2. Output\" section in <b>SNPcaster.ipynb</b>.<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d2b600c",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "## 5. Create Phylogenetic Tree\n",
    "\n",
    "Use the obtained final_snp.fasta etc. to perform phylogenetic tree construction using the maximum likelihood method (raxml-ng.sh).<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8311d1b0",
   "metadata": {},
   "source": [
    "### 5.1. Run RAxML-NG\n",
    "  - In the cell below, enter the path to the SNP file to be analyzed after \"input=\".\n",
    "  - Example: 'snpcaster_20240115_094704_list.txt/3_results_without_gubbins/final_snp.fasta'<br>\n",
    "    ↑In the above example, note that <u><font color=\"Red\">the \"20240115_094704_list.txt\" part differs for each analysis</u></font>.<br>\n",
    "      Also note that <u><font color=\"Red\">the folder name includes \".txt\"</u></font> as in snpcaster_20240115_094704_list<font color=\"Red\">.txt</font>/...\n",
    "    \n",
    "  - If necessary, change or delete the name of the reference sequence (default is \"Ref\").\n",
    "  - Note: Due to RAxML-NG specifications, <u><font color=\"Red\">an error will occur if there are 3 or fewer samples.</font></u>\n",
    "    - If you want to create a phylogenetic tree with 3 strains, use [11.1.2. IQTREE in SNPcaster.ipynb](./SNPcaster.ipynb#iqtree-run).\n",
    "\n",
    "\n",
    "- Command\n",
    "  - bash snpcaster_20230821/raxml-ng.sh [input] [threads] [bootstrap]\n",
    "<br>\n",
    "- Options\n",
    "\n",
    "|Parameter|Required|Description|\n",
    "|---|---|---|\n",
    "|input|●|Input file (e.g., xxx.fasta). Input alignment in PHYLIP/FASTA/NEXUS/CLUSTAL/MSF format|\n",
    "|threads|●|Number of modeltest-ng threads (e.g., 12) * raxml-ng threads are automatically selected|\n",
    "|bootstrap|-|bootstrap (default is 1000)|\n",
    "\n",
    "<br>\n",
    "\n",
    "<font color=\"Tomato\">────────────── ↓↓↓ ***Execute Command*** ↓↓↓ ──────────────</font><br>\n",
    "\\* For log information, refer to `raxml-ng.log` in the execution directory. （<u><font color=\"Red\">***This file is overwritten each time you run.***</u></font>）"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d77cb43d",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################\n",
    "# Set Parameters\n",
    "####################################################\n",
    "input = 'snpcaster_20240115_094704_list.txt/3_results_without_gubbins/final_snp.fasta'\n",
    "# To use results from running Gubbins, use the following line:\n",
    "# Delete the above input line => Remove the # at the beginning of the line below => Change 'snpcaster_20240115_094704_list.txt' to the SNPcaster result folder name.\n",
    "# input = \"snpcaster_20240115_094704_list.txt/5_results_with_gubbins/final_snp_after_gubbins.fasta\"\n",
    "threads = 8\n",
    "bootstrap = 1000\n",
    "\n",
    "####################################################\n",
    "# Run raxml-ng\n",
    "####################################################\n",
    "!bash raxml-ng.sh $input $threads $bootstrap > raxml-ng.log\n",
    "!echo 'Complete!'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63f83187",
   "metadata": {},
   "source": [
    "<font color=\"Tomato\">────────────────────────────────────────────</font>\n",
    "<br>\n",
    "\n",
    "### 5.2. Output Location for Phylogenetic Tree File\n",
    "  - A RAxML folder (raxml\\_results\\_[date]\\_[time]) will be created in the analysis folder, and the phylogenetic tree file will be output there.\n",
    "  - The `final_snp.fasta_bootstrap.nwk` in that folder is the phylogenetic tree file with bootstrap values.\n",
    "<br>\n",
    "\n",
    "\n",
    "### 5.3. Verify Phylogenetic Tree File\n",
    "#### 5.3.1. Maximum Likelihood Phylogenetic Tree\n",
    "Open the phylogenetic tree file with MEGA, CLC Genomics Workbench, etc.<br>\n",
    "To save images in MEGA, use Image/Copy to Clipboard. To save images with Root in MEGA, select the line that will be the Root, click \"Place Root on Branch\", and use \"Show Subtree Separately\".<br>\n",
    "<br>\n",
    "#### 5.3.2. Haplotype Network Diagram\n",
    "Use PopART to create haplotype network diagrams such as minimum spanning trees.\n",
    "Select and open final_snp.nex from the snpcaster_date_time_list_test/3_results_without_gubbins folder.<br>\n",
    "\\* This file is created as a result of snpcaster, so you don't need to run RAxML-NG."
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
